### ISSUE TITLE
denoising_diffusion_pytorch.py doesn't use eval mode for testing

### ISSUE THREAD

**Issue by samuelemarro** [lucidrains/denoising-diffusion-pytorch#327]
When computing the FID in `denoising_diffusion_pytorch.py` (as well as `denoising_diffusion_pytorch_1d.py`), the Trainer never switches the model to eval mode. This isn't an issue in the default implementation (since the UNet doesn't use any training-dependant modules, such as BatchNorm), but when modifying the architecture it can lead to unexpected problems.

Note that, despite its name, `torch.inference_mode()`, used in both the [training function](https://github.com/lucidrains/denoising-diffusion-pytorch/blob/3343eeb0a0f8cb78926dc0cc616c4f980537e264/denoising_diffusion_pytorch/denoising_diffusion_pytorch.py#L1063) and [as a decorator for some methods](https://github.com/lucidrains/denoising-diffusion-pytorch/blob/3343eeb0a0f8cb78926dc0cc616c4f980537e264/denoising_diffusion_pytorch/denoising_diffusion_pytorch.py#L736), doesn't set the model to eval mode. For a simple test, run:
```python
import torch
import torch.nn as nn

class SampleModule(nn.Module):
    def forward(self, x):
        print(self.training)
        return x

sample_module = SampleModule()

with torch.inference_mode():
    sample_module(torch.randn(3, 3))
```
which will print `True`.

**Comment by lucidrains** [comment]
@samuelemarro hey Samuele, thanks for raising this issue

i think this is taken care of [here](https://github.com/lucidrains/denoising-diffusion-pytorch/blob/main/denoising_diffusion_pytorch/fid_evaluation.py#L95)? but i just realized it may not set it back to train mode, which may be necessary given the dropouts just added

let me quickly do that

**Comment by lucidrains** [comment]
@samuelemarro oh actually, it is fine, since the `FidEvaluator` is working off the `EMA` model

**Comment by samuelemarro** [comment]
@lucidrains I've just finished looking over the EMA library. So, if I got it right, this is what happens:
- During `Trainer.__init__`, you initialize the `EMA`. This creates a deep copy of the model which is then passed to the `FidEvaluator`
- At testing time, you copy the parameters to the model stored in the `EMA`
- You call `.eval()` for the `FidEvaluator` model (which is actually the EMA model), which means that the original model isn't affected and remains in training mode

If this is correct, I'll go ahead and close the issue, If there isn't any downside, I'll also probably edit the script so that it calls `.eval()` and `.train()` manually (I need to compute some other metrics besides the FID).

Out of curiosity, what is the rationale for keeping a deep-copied EMA model and manually updating the parameters? Is it for the sake of thread safety / accelerator black magic?

**Comment by lucidrains** [comment]
@samuelemarro i'm not sure where the practice of keeping an EMA of generator came from, but i first encountered it in StyleGAN2 from Tero Karras years ago. ever since, many papers use this technique and it has become mainstream in usage
